core/trino-parser/src/main/java/io/trino/sql/parser/ErrorHandler.java

```diff
@@ -211,14 +211,11 @@ public Result process(ATNState currentState, int tokenIndex, RuleContext context
             }
 
             Set<Integer> endTokens = process(new ParsingState(currentState, tokenIndex, false, parser), 0);
-            Set<Integer> nextTokens = new HashSet<>();
             while (!endTokens.isEmpty() && context.invokingState != -1) {
-                for (int endToken : endTokens) {
-                    ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
-                    nextTokens.addAll(process(new ParsingState(nextState, endToken, false, parser), 0));
-                }
+                final ATNState nextState = ((RuleTransition) atn.states.get(context.invokingState).transition(0)).followState;
+                endTokens = endTokens.stream().flatMap(endToken ->
```

Comment:
Currently the trino-parser test fails but I'm not sure if the test itself is wrong.

Meaning, I can change line 216 to mimic the old behavior to give us the expected result for the test:
```
                endTokens = Stream.concat(endTokens.stream(), endTokens.stream().flatMap(endToken ->
                    process(new ParsingState(nextState, endToken, false, parser), 0).stream())).collect(Collectors.toSet());
```

The test in question that fails is:
`io.trino.sql.parser.TestSqlParserErrorHandling.testStatement(String, String)[46]`

Where the expected output is:
```
<"line 1:21: mismatched input 'SELECT'. Expecting: '%', '(', ')', '*', '+', ',', '-', '.', '/', 'AND', 'AS', 'AT', 'EXCEPT', 'FETCH', 'FROM', 'GROUP', 'HAVING', 'INTERSECT', 'LIMIT', 'OFFSET', 'OR', 'ORDER', 'SELECT', 'TABLE', 'UNION', 'VALUES', 'WHERE', 'WINDOW', '[', '||', <EOF>, <identifier>, <predicate>">
```

And where my current output is:
```
<"line 1:21: mismatched input 'SELECT'. Expecting: '%', ')', '*', '+', ',', '-', '.', '/', 'AND', 'AS', 'AT', 'EXCEPT', 'FETCH', 'FROM', 'GROUP', 'HAVING', 'INTERSECT', 'LIMIT', 'OFFSET', 'OR', 'ORDER', 'UNION', 'WHERE', 'WINDOW', '[', '||', <identifier>, <predicate>">
```

The diff looks like this:
![test_diff](https://user-images.githubusercontent.com/136857/217172932-3f9840af-321d-406d-a043-f97eaf20fbee.png)